### PR TITLE
Increase step between checks for cancellation

### LINKED ISSUE #53496: ExitableDirectoryReader slowdowns intensive points and terms query
The exitable directory reader added in https://github.com/elastic/elasticsearch/pull/52822 seems to slowdown queries that uses points (`range` queries) and terms dictionary (`terms` ad multi-term queries) intensively. These slowdowns have been spotted in our nightly runs as outlined [here](https://elasticsearch-benchmarks.elastic.co/#tracks/noaa/nightly/default/30d). 
We open this issue to discuss the best ways to mitigate these slowdowns. The first thing we'd like to try is to make the interval at which we check for cancellation or timeout smaller since we started with a relatively low value (15).

### PR DESCRIPTION
**Pull request by matriv**
The introduction of the ExitableDirectoryReader showed increase of 
latencies for range queries using pointvalues. 

Check for cancellation every 1024 docs instead of every 15 to lower
the impact of the check in query's performance.

Follows: #52822
Fixes: #53496

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java [hunk 1]
```diff
@@ -245,7 +245,7 @@ public int getDocCount() {
 
     private static class ExitableIntersectVisitor implements PointValues.IntersectVisitor {
 
-        private static final int MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK = (1 << 4) - 1; // 15
+        private static final int MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK = 1 << 10; // 1024
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java]:
```suggestion
        private static final int MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK = (1 << 10) - 1; // 1023
```

**Review by jimczi — APPROVED** [review]
I left one comment that needs to be addressed

**Comment by matriv** [comment]
Benchmark results showing the effect of the change:
```
      50th percentile latency |    range |   799.665 |   632.712 | -166.953 |  ms
      90th percentile latency |    range |   819.129 |   653.389 | -165.741 |  ms
      99th percentile latency |    range |   824.655 |   670.564 |  -154.09 |  ms
     100th percentile latency |    range |   824.683 |   675.369 | -149.314 |  ms
 50th percentile service time |    range |   799.465 |   632.326 | -167.139 |  ms
 90th percentile service time |    range |   818.915 |   653.021 | -165.894 |  ms
 99th percentile service time |    range |   824.430 |   670.197 | -154.233 |  ms
100th percentile service time |    range |   824.464 |   675.002 | -149.461 |  ms
```

**Comment by matriv** [comment]
Here are the results between the status before the ExitableDirectoryReader and after but with the change to check every 1023 docs instead of every 15:
```
      50th percentile latency |  range |  641.796 |  642.785 |   0.98898
      90th percentile latency |  range |  643.440 |  649.266 |   5.82666
      99th percentile latency |  range |  650.121 |  692.947 |  42.82630
     100th percentile latency |  range |  650.332 |  712.341 |  62.00830
 50th percentile service time |  range |  641.418 |  642.400 |   0.98233
 90th percentile service time |  range |  643.063 |  648.894 |   5.83093
 99th percentile service time |  range |  649.741 |  692.553 |  42.81110
100th percentile service time |  range |  649.957 |  711.938 |  61.98130
```
FYI @jimczi

**Pull request merged**